engine/common/insight_log/resource_insight_logs.py

```diff
@@ -0,0 +1,132 @@
+import enum
+import json
+import logging
+import re
+from abc import ABC, abstractmethod
+
+from django.apps import apps
+
+insight_logger = logging.getLogger("insight_logger")
+logger = logging.getLogger(__name__)
+
+
+class EntityEvent(enum.Enum):
+    CREATED = "created"
+    UPDATED = "updated"
+    DELETED = "deleted"
+
+
+class InsightLoggable(ABC):
+    @property
+    @abstractmethod
+    def public_primary_key(self):
+        pass
+
+    @property
+    @abstractmethod
+    def insight_logs_verbal(self) -> str:
+        """
+        insight_logs_verbal returns resource name for insight_log
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def insight_logs_type_verbal(self) -> str:
+        """
+        insight_logs_type_verbal resource type for insight_log
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def insight_logs_serialized(self) -> dict:
+        """
+        insight_logs_serialized returns resource, serialized for insight_log
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def insight_logs_metadata(self) -> dict:
+        """
+        insight_logs_metadata returns resource metadata which should be present in the insight_log line
+        """
+        pass
+
+
+def resource_insight_log(instance: InsightLoggable, author, event: EntityEvent, prev_state=None, new_state=None):
+    try:
+        organization = author.organization
+        DynamicSetting = apps.get_model("base", "DynamicSetting")
+        org_id_to_enable_insight_logs, _ = DynamicSetting.objects.get_or_create(
+            name="org_id_to_enable_insight_logs",
+            defaults={"json_value": []},
+        )
+        log_all = "all" in org_id_to_enable_insight_logs.json_value
+        insight_logs_enabled = organization.id in org_id_to_enable_insight_logs.json_value
+        if insight_logs_enabled or log_all:
+            tenant_id = organization.stack_id
+            author_id = author.public_primary_key
+            author = json.dumps(author.username)
+            entity_type = instance.insight_logs_type_verbal
+            try:
+                entity_id = instance.public_primary_key
+            except AttributeError:
+                entity_id = instance.id  # Fallback for entities which have no public_primary_key
```

Comment:
does it mean internal id's are exposed to the customer? 
can we just use None or - here
I suppose it is not used in the log routing in our internal logic, only for the customers

Reply:
It will be exposed only for api access tokens. And for them it's already exposed:
<img width="1840" alt="image" src="https://user-images.githubusercontent.com/20221722/186102473-f590ab45-26fa-4ccd-9942-ea3e49a14d7a.png">

Reply:
can you write the comment that fallback is for other types of instances. Because now it looks like some objects have the id, but others don't within the same type.